### PR TITLE
Feature/2 users authentication

### PR DESCRIPTION
**Pull request by lisset-orozco**
- [x] create User model, configure shoulda-matchers and user_spec

- [x] token model

- [x] token test

- [x] namespaces for api

- [x] user_provider_uid from_omniauth

- [x] user login

- [x] UserController request test

- [x] authenticate users

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,19 @@
 class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
-  protect_from_forgery with: :exception
+  # protect_from_forgery with: :exception
+  protect_from_forgery with: :null_session
+
+  before_action :authenticate
+
+  def authenticate
```
Comment (joalbertg) [inline app/controllers/application_controller.rb]: refactor code! create memorize method

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+    token_str = params[:token]
+    token = Token.find_by(token: token_str)
+
+    if token.nil? || !token.active?
```
Comment (joalbertg) [inline app/controllers/application_controller.rb]: change to return if/unless

File: app/models/token.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# token
+class Token < ActiveRecord::Base
+  # -- Relationships --------------------------------------------------------
+  belongs_to :user
+  before_create :generate_token
```
Comment (joalbertg) [inline app/models/token.rb]: refactor to service class

File: app/views/api/v1/users/show.json.jbuilder [hunk 4]
```diff
@@ -0,0 +1,3 @@
+json.(@user, :id, :name, :email)
+
+json.(@token, :token)
```
Comment (joalbertg) [inline app/views/api/v1/users/show.json.jbuilder]: Add line at end of file

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (lisset-orozco) [reply app/controllers/application_controller.rb]: Done!

File: app/views/api/v1/users/show.json.jbuilder (hunk 4 above)
Comment (lisset-orozco) [reply app/views/api/v1/users/show.json.jbuilder]: done!

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (lisset-orozco) [reply app/controllers/application_controller.rb]: refactor without memoized

File: app/models/token.rb (hunk 3 above)
Comment (lisset-orozco) [reply app/models/token.rb]: done

File: app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+module Api
+  module V1
+    # users
+    class UsersController < ApiV1Controller
+      # POST /users
+      def create
+        # params = { auth: { provider: , uid: } }
+        return error_message('param', :auth) if params[:auth]
+
+        @user = User.from_omniauth(params.fetch(:auth))
```
Comment (joalbertg) [inline app/controllers/api/v1/users_controller.rb]: refactor equal code

File: app/controllers/api/v1/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        @token = @user.tokens.create
+
+        render 'api/v1/users/show'
+      end
+
+      def show
+        # params = { auth: { provider: , uid: } }
+        return error_message('param', :auth) if params[:auth].nil?
```
Comment (joalbertg) [inline app/controllers/api/v1/users_controller.rb]: redundant nil? method.

File: app/controllers/api_v1_controller.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  include ErrorHandler
+
+  before_action :authenticate
+
+  protected
+
+  def authenticate
+    return error_message('token', :unauthorized) unless Token.find_token(params[:token])
+
+    @current_user = Token.token_data.user
```
Comment (joalbertg) [inline app/controllers/api_v1_controller.rb]: refactor to .user

File: app/controllers/concerns/error_handler.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+module ErrorHandler
+  extend ActiveSupport::Concern
+
+  MSG = { param: 'param is missing',
+          token: 'Invalid token' }.freeze
+
+  def error_message(type, *value)
+    # value = param_name || status
+
+    message = MSG[type.to_sym]
```
Comment (joalbertg) [inline app/controllers/concerns/error_handler.rb]: one method!

File: app/controllers/concerns/error_handler.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    message = "#{value.first} " + message if type.eql?('param')
+    message = { error: message }
```
Comment (joalbertg) [inline app/controllers/concerns/error_handler.rb]: two method!

File: app/models/token.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  belongs_to :user
+
+  def active?
+    expires_at > DateTime.now
+  end
+
+  class << self
+    attr_reader :token_data
+
+    def find_token(token_str)
```
Comment (joalbertg) [inline app/models/token.rb]: change to token? method

File: app/models/token.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      @token_data = Token.find_by(token: token_str) unless token_str.nil?
+      @token_data if !@token_data.nil? && @token_data.active?
```
Comment (joalbertg) [inline app/models/token.rb]: return `true` or `false`

File: app/views/api/v1/users/show.json.jbuilder (hunk 4 above)
Comment (joalbertg) [reply app/views/api/v1/users/show.json.jbuilder]: Add line at end of file

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/views/api/v1/users/show.json.jbuilder (hunk 4 above)
Comment (lisset-orozco) [reply app/views/api/v1/users/show.json.jbuilder]: done

File: app/controllers/api_v1_controller.rb (hunk 7 above)
Comment (lisset-orozco) [reply app/controllers/api_v1_controller.rb]: done

File: app/controllers/api/v1/users_controller.rb (hunk 5 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/users_controller.rb]: done

File: app/controllers/api/v1/users_controller.rb (hunk 6 above)
Comment (lisset-orozco) [reply app/controllers/api/v1/users_controller.rb]: done

File: app/models/token.rb (hunk 10 above)
Comment (lisset-orozco) [reply app/models/token.rb]: done

File: app/models/token.rb (hunk 11 above)
Comment (lisset-orozco) [reply app/models/token.rb]: done

File: app/controllers/concerns/error_handler.rb (hunk 9 above)
Comment (lisset-orozco) [reply app/controllers/concerns/error_handler.rb]: done

File: app/controllers/concerns/error_handler.rb (hunk 8 above)
Comment (lisset-orozco) [reply app/controllers/concerns/error_handler.rb]: done

**Review by joalbertg — APPROVED** [review]

**Pull request merged**